sql/sql_audit.cc

```diff
@@ -1311,6 +1316,13 @@ static int event_class_dispatch(THD *thd, mysql_event_class_t event_class,
                ? 1
                : 0;
   } else {
+    // mutexes are non-reentrant, and audit_null dispatches events from this
+    // function
+    thread_local bool mutex_locked = false;
```

Comment:
Although, this might work, it looks kinda hackish.
Did you manage to figure out the original source of recursive `event_class_dispatch()` calls?

Reply:
It only happens in audit null and mtr: it has a feature that creates (user specified) events on an event dispatch.

Reply:
Please consider marking events generated from inside `audit_null` plugin with a special flag (say, `GENERATED_FROM_DISPATCH`) and lock `thd->LOCK_thd_audit_data` depending on the value of this flag.

If this is too much trouble and / or requires significant audit interface changes, then `thread_local` solution is fine.

Reply:
I tried a few approaches, but I think this would end up more complex/hackish than this. While this bool isn't the best, it only adds 6 lines of code to the patch, is a localized change, and quite easy to understand.

Reply:
OK, then it's good as is.

Reply:
Actually, any audit plugin that calls my_message (or generates errors) will end up recursively calling the dispatch function. I don't think this will interact well if we have different THDs on different physical threads.

Since the `audit_class_plugins` list is append-only, and plugins are never free'd until the THD is done, I think a possible solution is just to copy the list of plugins under the lock, but iterate through the list while unlocked.

Reply:
I wanted to avoid possible additional allocations for every notification - but I guess if you are using less audit plugins than the preallocated size for the container, that won't be an issue there.